optuna/pruners/threshold.py

```diff
@@ -0,0 +1,140 @@
+        upper:
+            A maximum value which determines whether pruner prunes or not.
+            If an intermediate value is larger than upper, it prunes.
+        n_warmup_steps:
+            Pruning is disabled until the trial exceeds the given number of step.
+        interval_steps:
+            Interval in number of steps between the pruning checks, offset by the warmup steps.
+            If no value has been reported at the time of a pruning check, that particular check
+            will be postponed until a value is reported. Value must be at least 1.
+
+    """
+
+    def __init__(
+        self,
+        lower: Optional[float] = None,
+        upper: Optional[float] = None,
+        n_warmup_steps: int = 0,
+        interval_steps: int = 1,
+    ) -> None:
+
+        if lower is not None and not isinstance(lower, float):
+            raise TypeError("lower should be either None or a floating point.")
+        if upper is not None and not isinstance(upper, float):
+            raise TypeError("upper should be either None or a floating point.")
+        if lower is None and upper is None:
+            raise ValueError("Either lower or upper must be specified.")
+
+        lower = lower if lower is not None else -float("inf")
+        upper = upper if upper is not None else float("inf")
+
+        if lower > upper:
+            raise ValueError("lower should be smaller than upper.")
+        if n_warmup_steps < 0:
+            raise ValueError(
+                "Number of warmup steps cannot be negative but got {}.".format(n_warmup_steps)
+            )
+        if interval_steps < 1:
+            raise ValueError(
+                "Pruning interval steps must be at least 1 but got {}.".format(interval_steps)
+            )
+
+        self._lower = lower
+        self._upper = upper
+        self._n_warmup_steps = n_warmup_steps
+        self._interval_steps = interval_steps
+
+    def prune(self, study: "optuna.study.Study", trial: optuna.structs.FrozenTrial) -> bool:
+
+        step = trial.last_step
+        if step is None:
+            return False
+
+        n_warmup_steps = self._n_warmup_steps
+        if step <= n_warmup_steps:
+            return False
+
+        if not _is_first_in_interval_step(
+            step, trial.intermediate_values.keys(), n_warmup_steps, self._interval_steps
+        ):
+            return False
+
+        latest_value = trial.intermediate_values[step]
+        if math.isnan(latest_value):
+            return True
+
+        if latest_value < self._lower:
+            return True
+
+        if latest_value > self._upper:
+            return True
+
+        return False
+
+    @property
+    def lower(self) -> Optional[float]:
+
+        if self._lower == -float("inf"):
+            return None
+        return self._lower
+
+    @property
+    def upper(self) -> Optional[float]:
+
+        if self._upper == float("inf"):
+            return None
+        return self._upper
```

Comment:
These properties will return `None` when the user passes `float("inf")` or `-float("inf")` to `__init__` which might be confusing. I bet it's safer to simply hide these properties as a first implementation. What do you think?

Reply:
It makes sense and I didn't consider the case user passes them directly.
Thank you for the suggestion! (hired in 6f710802)